### ISSUE TITLE
Shutdown hooks for factories

### ISSUE THREAD

**Issue by maguro** [envoyproxy/envoy#27990]
It would be nice if factories had a mechanism whereby they would be notified when the Envoy server is shutting down.

**Comment by kyessenov** [comment]
Is [lifecycle notifier](https://github.com/envoyproxy/envoy/blob/main/envoy/server/lifecycle_notifier.h) not enough?

**Comment by maguro** [comment]
Maybe @doujiang24 may be able to answer?

**Comment by doujiang24** [comment]
> Is this exposed to golang filters and extensions?

@maguro Not yet. But it could be possible. Could you please describe the use case a bit? Thanks.

**Comment by maguro** [comment]
@doujiang24 , I was thinking

    type StreamFilterFactoryV2 func(ctx context.Context, callbacks FilterCallbackHandler) StreamFilter

would be the most idiomatic.

**Comment by doujiang24** [comment]
Thanks @maguro , I see that you want a hook in Go side when the Envoy server is shutting down.
Why do you need the hook? for what kind of case?

**Comment by maguro** [comment]
@doujiang24 , that's correct.  I had to create a bunch of stuff in `init()` that I'd like to tidy up and close when the plugin is unloaded.  As I think on it, I think this would be more useful

    type StreamFilterConfigFactoryV2 func(ctx context.Context, config interface{}) StreamFilterFactory

**Comment by doujiang24** [comment]
Oh, I see, that make sense.
It also could be possible, I think we could call into Go in the Dso instance destruct method: https://github.com/envoyproxy/envoy/blob/main/contrib/golang/common/dso/dso.cc#L37

@maguro Do you want to have a try? I'm happy to review.

**Comment by maguro** [comment]
Sure.

**Comment by doujiang24** [comment]
Oh, I think it's useful, please mark it no stale, we could implement it in the feature. @phlax Thanks
